File: android/src/main/java/viska/android/NewProfileActivity.java
package viska.android;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import com.couchbase.lite.Database;
import com.uber.autodispose.AutoDispose;
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider;
import io.reactivex.Completable;
import io.reactivex.schedulers.Schedulers;
import viska.database.DatabaseKt;

public class NewProfileActivity extends AppCompatActivity {

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.new_profile);

    final Button newMockProfileButton = findViewById(R.id.new_mock_profile);
    newMockProfileButton.setOnClickListener(view -> onNewMockProfile());

    final Application app = (Application) getApplication();
    final ProgressBar progressBar = findViewById(R.id.progress);
    final Button newAccountButton = findViewById(R.id.new_account);
    app.getViewModel()
        .creatingAccount
        .observe(
            this,
            running -> {
              if (running) {
                progressBar.setVisibility(View.VISIBLE);
                newAccountButton.setVisibility(View.GONE);
                newMockProfileButton.setVisibility(View.GONE);
              } else {
                progressBar.setVisibility(View.GONE);
                newAccountButton.setVisibility(View.VISIBLE);
                newMockProfileButton.setVisibility(BuildConfig.DEBUG ? View.VISIBLE : View.GONE);
              }
            });
  }

  private void onNewMockProfile() {
    final Application app = (Application) getApplication();
    app.getViewModel().creatingAccount.setValue(true);
    Completable.fromAction(
            () -> {
              final Database db = DatabaseKt.open();
              DatabaseKt.createDemoProfile(db);
              db.close();
              app.getViewModel().creatingAccount.postValue(false);
            })
        .observeOn(Schedulers.io())
        .subscribeOn(Schedulers.from(getMainExecutor()))
        .as(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(this)))
        .subscribe(
            () -> {
              startActivity(new Intent(this, MainActivity.class));
              finish();
            });
  }
}

File: android/src/main/java/viska/android/NewProfileActivity.kt
package viska.android

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material.Button
import androidx.compose.material.LinearProgressIndicator
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.setContent
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import dagger.Lazy
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import viska.daemon.DaemonService
import viska.database.ProfileService

@AndroidEntryPoint
class NewProfileActivity : AppCompatActivity() {

  @Inject lateinit var profileService: ProfileService
  @Inject lateinit var daemonService: Lazy<DaemonService>

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContent {
      val creatingAccount by GlobalState.creatingAccount.collectAsState()
      Page(creatingAccount, { createProfile(false) }, { createProfile(true) })
    }
  }

  private fun createProfile(mock: Boolean) {
    GlobalState.creatingAccount.value = true
    InstanceActivity.finishAll()

    GlobalScope.launch(Dispatchers.IO) {
      try {
        profileService.createProfile(mock)
      } finally {
        GlobalState.creatingAccount.value = false
        startActivity(Intent(this@NewProfileActivity, DashboardActivity::class.java))
        finish()
      }
    }
  }
}

@Composable
private fun Page(creatingAccount: Boolean, newAccount: () -> Unit, newMockProfile: () -> Unit) {
  Column(
      verticalArrangement = Arrangement.SpaceEvenly,
      horizontalAlignment = Alignment.CenterHorizontally,
      modifier = Modifier.fillMaxSize(),
  ) {
    Button(onClick = newAccount, enabled = !creatingAccount) {
      Text(stringResource(R.string.new_account))
    }

    if (BuildConfig.DEBUG) {
      Button(onClick = newMockProfile, enabled = !creatingAccount) {
        Text(stringResource(R.string.new_mock_profile))
      }
    }

    if (creatingAccount) {
      LinearProgressIndicator()
    }
  }
}

@Preview
@Composable
private fun CreatingAccount() {
  Page(true, {}, {})
}

@Preview
@Composable
private fun NotCreatingAccount() {
  Page(false, {}, {})
}
